File: dotnet/ThirtyFour.UI/Table/Table.cs
using System.Collections.Generic;
using OpenQA.Selenium;

namespace ThirtyFour.UI.Table
{
    /// <summary>
    /// The Table class provides helper methods for interacting with a HTML table
    /// </summary>
    public class Table : WrappedElement
    {

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="element">The root element of the table</param>
        public Table(IWebElement element) : base(element)
        {
        }

        /// <summary>
        /// Returns all header cells
        /// </summary>
        public IReadOnlyCollection<IWebElement> HeaderCells
        {
            get
            {
                return Element.FindElements(By.TagName("th"));
            }
        }

        /// <summary>
        /// Returns the body rows
        /// </summary>
        public IReadOnlyCollection<IWebElement> BodyRows
        {
            get
            {
                return Element.FindElements(By.CssSelector("tbody tr"));
            }

        }
        /// <summary>
        /// Returns a cell value as a String by correlating the table headers with a cell value
        /// </summary>
        /// <param name="findColumn">The column header used to identify the correct table row</param>
        /// <param name="findValue">The cell value used to identify the correct table row</param>
        /// <param name="returnColumn">The column header used to identify the return cell</param>
        /// <returns>The value of the cell if found otherwise a NoSuchElementException will be thrown</returns>
        public string GetCellValue(string findColumn, string findValue, string returnColumn)
        {
            ICellFinder finder = new ColumnHeaderCellFinder(findColumn, findValue, returnColumn);
            return this.GetCellValue(finder, new StringCellParser());
        }

        /// <summary>
        /// Returns a cell value based on custom ICellFinder and ICellParser strategies
        /// </summary>
        /// <typeparam name="T">The data type for the generics cell parser</typeparam>
        /// <param name="cellFinder"></param>
        /// <param name="cellParser"></param>
        /// <returns></returns>
        public T GetCellValue<T>(ICellFinder cellFinder, ICellParser<T> cellParser)
        {
            IWebElement cell = cellFinder.FindCell(Element);
            return cellParser.GetValue(cell);
        }
    }
}

File: java/src/main/java/com/github/danhumphrey/thirtyfour/ui/table/Table.java
package com.github.danhumphrey.thirtyfour.ui.table;

import java.util.List;

import org.openqa.selenium.By;
import org.openqa.selenium.NoSuchElementException;
import org.openqa.selenium.WebElement;

import com.github.danhumphrey.thirtyfour.ui.WrappedElement;

/**
 * The Table class provides helper methods for interacting with a HTML table
 * @author dhumphrey
 *
 */
public class Table extends WrappedElement{

	/**
	 * Constructor
	 * @param element the root element of the table
	 */
	public Table(WebElement element) {
		super(element);
	}
	
	/**
	 * Returns a list of the table header cells
	 * @return a list of WebElement instances representing the th elements
	 */
	public List<WebElement> getHeaderCells() {
		return this.element.findElements(By.tagName("th"));
	}
	
	/**
	 * Returns a list of the body rows 
	 * @return a list of WebElement instances representing the tr elements
	 */
	public List<WebElement> getBodyRows() {
		return this.element.findElements(By.cssSelector("tbody tr"));
	}
	
	/**
	 * Returns a cell value as a String by correlating the table headers with a cell value
	 * @param findColumn The column header used to identify the correct table row
	 * @param findValue The cell value used to identify the correct table row
	 * @param returnColumn The column header used to identify the return cell
	 * @return the value of the cell
	 * @throws NoSuchElementException if the column or row can't be found
	 * @see ColumnHeaderCellFinder
	 * {@link Table#getCellValue(CellFinder, CellParser)}
	 */
	public String getCellValue(String findColumn, String findValue, String returnColumn) throws NoSuchElementException {
		CellFinder finder = new ColumnHeaderCellFinder(findColumn, findValue, returnColumn);
		return this.getCellValue(finder, new StringCellParser()); 
	}
	
	/**
	 * Returns a cell value by the data type specified in the generics cellParser
	 * @param <T> the type returned by the cell parser
	 * @param cellFinder the strategy used to find the correct cell
	 * @param cellParser a generics cell parser used to extract the value from the cell
	 * @return the value of the cell
	 * @throws NoSuchElementException if the column or row can't be found
	 * {@link Table#getCellValue(String, String, String)}
	 */
	public <T> T getCellValue(CellFinder cellFinder, CellParser<T> cellParser) throws NoSuchElementException{
		WebElement cell = cellFinder.findCell(this.element);
		return cellParser.getValue(cell);
	}	
}